cluster_experiments/experiment_analysis.py

```diff
@@ -680,3 +682,255 @@ def analysis_point_estimate(self, df: pd.DataFrame, verbose: bool = False) -> fl
+            .T.iloc[:, 0]
+        )
+
+        weights = get_w(X, y, verbose)
+
+        return weights
+
+    def synthetic_control(
+        self,
+        pre_experiment_df: pd.DataFrame,
+        df: pd.DataFrame,
+        treatment_cluster,
+        verbose: bool = False,
+    ) -> pd.DataFrame:
+        """
+        This method adds a column with the synthetic results and filter only the treatment unit.
+
+        First, it calculates the weights of each donor in the control group using the `fit_synthetic` method.
+        It then uses these weights to create a synthetic control group that closely matches the treatment unit before the intervention.
+        The synthetic control group is added to the treatment unit in the dataframe.
+
+        Args:
+            pre_experiment_df (pd.DataFrame): The dataframe containing the data before the intervention.
+            df (pd.DataFrame): The dataframe containing the data after the intervention.
+            treatment_cluster (str): The name of the treatment cluster.
+
+        Returns:
+            pd.DataFrame: The dataframe with the synthetic results added to the treatment cluster.
+        """
+        weights = self.fit_synthetic(
+            pre_experiment_df=pre_experiment_df, verbose=verbose
+        )
+
+        synthetic = (
+            df.query(
+                f"{''.join(self.cluster_cols)} != '{treatment_cluster}'"
+            )  # remove treatment cluster
+            .pivot(index=self.time_col, columns=self.cluster_cols)[self.target_col]
+            .values.dot(weights)
+        )
+
+        return df.query(
+            f"{''.join(self.cluster_cols)} == '{treatment_cluster}'"
+        ).assign(  # add synthetic to treatment cluster
+            synthetic=synthetic
+        )
+
+    def calculate_effects(self, synthetic_donors: List[pd.DataFrame]):
+        """
+        Calculate the average treatment effect and the placebo effects.
+
+        The method identifies the treatment unit and separates its average effect (the real treatment effect) from
+        the average effects of the other units (the placebo effects).
+
+        Args:
+            synthetic_donors (list): A list of dataframes, each representing a cluster
+
+        Returns:
+            tuple: A tuple containing the average treatment effect and a list of the placebo effects.
+        """
+        synthetic_donors = [
+            unit.assign(effect=unit[self.target_col] - unit["synthetic"])
+            for unit in synthetic_donors
+        ]
+
+        avg_effects = [
+            unit.query(f"{self.time_col} >= '{self.intervention_date}'")[
+                "effect"
+            ].mean()
+            for unit in synthetic_donors
+        ]
+
+        treatment_index = np.where(
+            [df[self.treatment_col].any() for df in synthetic_donors]
+        )[0][0]
+        ate = avg_effects[treatment_index]  # this is the real treatment effect
```

Comment:
I think a more natural way to return this is a dictionary with {cluster: effect} pairs. wdyt?